File: Coder Army/SOLID/LSP(LiskovSubstitutionPrinciple)/LSP_followed.cpp
#include <bits/stdc++.h>
using namespace std;

class DepositOnlyAccount
{
public:
    virtual void deposit(int amount) = 0;
};
class Account : public DepositOnlyAccount
{
public:
    virtual void withdraw(int amount) = 0;
};

class SavingAccount : public Account
{
private:
    int balance;

public:
    SavingAccount() : balance(0) {}
    void deposit(int amount) override
    {
        balance += amount;
        cout << "Deposited: " << amount << " in the saving Account, New Balance: " << balance << endl;
    }

    void withdraw(int amount) override
    {
        if (balance >= amount)
        {
            balance -= amount;
            cout << "Withdrawn: " << amount << " from the saving Accouint. new Balance: " << balance << endl;
        }
        else
        {
            cout << "not enough balance in saving account\n";
        }
    }
};
class CurrentAccount : public Account
{
private:
    int balance;

public:
    CurrentAccount() : balance(0) {}
    void deposit(int amount) override
    {
        balance += amount;
        cout << "Deposited: " << amount << " in the current Account, New Balance: " << balance << endl;
    }

    void withdraw(int amount) override
    {
        if (balance >= amount)
        {
            balance -= amount;
            cout << "Withdrawn: " << amount << " from the current Accouint. new Balance: " << balance << endl;
        }
        else
        {
            cout << "not enough balance in current account\n";
        }
    }
};
class FixedAccount : public DepositOnlyAccount
{
private:
    int balance;

public:
    FixedAccount() : balance(0) {}
    void deposit(int amount) override
    {
        balance += amount;
        cout << "Deposited: " << amount << " in the fixed Account, New Balance: " << balance << endl;
    }
};

class BankClient
{
private:
    vector<DepositOnlyAccount *> depositAccounts;
    vector<Account *> withdrawlAccounts;

public:
    BankClient(vector<DepositOnlyAccount *> depositAccounts,
               vector<Account *> withdrawlAccounts) : depositAccounts(depositAccounts), withdrawlAccounts(withdrawlAccounts) {}

    void processTransactions()
    {
        for (Account *acc : withdrawlAccounts)
        {
            acc->deposit(1000);
            acc->withdraw(500);
        }
        for (DepositOnlyAccount *acc : depositAccounts)
        {
            acc->deposit(5000);
        }
    }
};

int main()
{
    vector<Account *> accounts;
    accounts.push_back(new SavingAccount());
    accounts.push_back(new CurrentAccount());

    vector<DepositOnlyAccount *> depositAccounts;
    depositAccounts.push_back(new FixedAccount());

    BankClient *client = new BankClient(depositAccounts, accounts);
    client->processTransactions();

    for (Account *acc : accounts)
    {
        delete acc;
    }
    for (DepositOnlyAccount *acc : depositAccounts)
    {
        delete acc;
    }
    delete client;
    return 0;
}

File: Coder Army/SOLID/LSP(LiskovSubstitutionPrinciple)/Javascript code/LSP_followed.js
class DepositOnlyAccount {
  deposit(amount) {
    throw new Error("deposit() must be implemented.");
  }
}

class Account extends DepositOnlyAccount {
  withdraw(amount) {
    throw new Error("withdraw() must be implemented.");
  }
}

class SavingAccount extends Account {
  constructor() {
    super();
    this.balance = 0;
  }

  deposit(amount) {
    this.balance += amount;
    console.log(`Deposited: ${amount} in the saving account, New Balance: ${this.balance}`);
  }

  withdraw(amount) {
    if (this.balance >= amount) {
      this.balance -= amount;
      console.log(`Withdrawn: ${amount} from the saving account. New Balance: ${this.balance}`);
    } else {
      console.log("Not enough balance in saving account");
    }
  }
}

class CurrentAccount extends Account {
  constructor() {
    super();
    this.balance = 0;
  }

  deposit(amount) {
    this.balance += amount;
    console.log(`Deposited: ${amount} in the current account, New Balance: ${this.balance}`);
  }

  withdraw(amount) {
    if (this.balance >= amount) {
      this.balance -= amount;
      console.log(`Withdrawn: ${amount} from the current account. New Balance: ${this.balance}`);
    } else {
      console.log("Not enough balance in current account");
    }
  }
}

class FixedAccount extends DepositOnlyAccount {
  constructor() {
    super();
    this.balance = 0;
  }

  deposit(amount) {
    this.balance += amount;
    console.log(`Deposited: ${amount} in the fixed account, New Balance: ${this.balance}`);
  }
}

class BankClient {
  constructor(depositAccounts, withdrawAccounts) {
    this.depositAccounts = depositAccounts;
    this.withdrawAccounts = withdrawAccounts;
  }

  processTransactions() {
    for (const acc of this.withdrawAccounts) {
      acc.deposit(100);
      acc.withdraw(40);
    }

    for (const acc of this.depositAccounts) {
      acc.deposit(500);
    }
  }
}

// Test
const withdrawAccounts = [
  new SavingAccount(),
  new CurrentAccount()
];

const depositAccounts = [
  new FixedAccount()
];

const client = new BankClient(depositAccounts, withdrawAccounts);
client.processTransactions();
